comptime.md

```diff
@@ -0,0 +1,151 @@
+# Compile-time execution
+
+# Summary 
+
+This RFC proposes to entirely change the current implementation of compile-time that will allow to run V language itself at comptime.
+
+# Motivation
+
+ The main purpose to do so is current limitation of compile-time evaluation:
+- No way to invoke arbitrary V functions
+- No way to choose function at compile time: 
+```v
+$if flag_a {
+    fn foo() {
+        println('bar')
+    }
+} $else {
+    fn foo() {
+        println('foo')
+    }
+}
+```
+^ Compiler cannot handle this code 
+
+- It limits our generics and compile-time reflection. With proper comptime it would be possible to generate complex serialization/deserialization functions without compiler support. 
+- Implementing proper comptime in fact would simplify our compiler pipelines, at the moment there is a lot of edge-cases for handling current limited comptime. 
+- Proper metaprogramming is must have in every modern language 
+
+
+# Guide-level explanation 
+
+With this RFC the main change is that all expressions and statements prefixed with `comptime` will be treated as compile-time and all of them can be executed *before* codegen. Current compile-time is quite limited and allows only a few statements (`$if`,`$for`) and only specific calls.
```

Comment:
Should we really change $ to comptime?

Reply:
No, this is about introducing `comptime` keyword so code using `$`-based comptime code does not break

Reply:
Hm, this would mean a really hard violation of the *there is only one way of doing a thing* principle. Why would you want to keep `$if` next to `comptime if` etc.?

Reply:
I think they co-exist some time but $if will be removed then.

Reply:
> I think they co-exist some time but $if will be removed then.

@dumblob this ^. Also we can emit error or warning when `$` is used in CTE context

Reply:
Sorry for the delay. I'm fine with this RFC, but I don't think we should replace `$` with `comptime`.

I'd keep `$`.

Reply:
> Sorry for the delay. I'm fine with this RFC, but I don't think we should replace `$` with `comptime`.
> 
> I'd keep `$`.

The point of adding `comptime` is to not break compiler while CTE is implemented so they would exist at the same time for some time and once comp time is stable enough just drop `comptime`

Reply:
Then it's totally fine :)